kernel/memory_structs/src/lib.rs

```diff
@@ -266,87 +315,145 @@ macro_rules! implement_page_frame {
 
             #[doc = "A `" $TypeName "` is a chunk of **" $desc "** memory aligned to a [`PAGE_SIZE`] boundary."]
             #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
-            pub struct $TypeName {
+            pub struct $TypeName<P: PageSize = Page4KiB> {
                 number: usize,
+                pub size: PhantomData::<P>,
             }
-
             impl $TypeName {
-                #[doc = "Returns the `" $address "` at the start of this `" $TypeName "`."]
-                pub const fn start_address(&self) -> $address {
-                    $address::new_canonical(self.number * PAGE_SIZE)
+                #[doc = "Returns the `" $TypeName "` containing the given `" $address "`."]
+                pub const fn containing_address(addr: $address) -> $TypeName {
+                    $TypeName {
+                        number: addr.value() / PAGE_SIZE,
+                        size: PhantomData::<Page4KiB>,
+                    }
                 }
-
+            }
+            impl<P> $TypeName<P> where P: PageSize + 'static {
```

Comment:
does the compiler require these `'static` lifetime annotations? If so, that's fine, but I see some places without them. Let's make it consistent across this entire file.